tests/integration/pilot/reachabiity/mtls_test.go

```diff
@@ -0,0 +1,285 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Verify reachability under different authN scenario.
+package reachability
+
+import (
+	"fmt"
+	"path"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/apps"
+	"istio.io/istio/pkg/test/framework/components/environment"
+	"istio.io/istio/pkg/test/framework/components/environment/kube"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	pilot2 "istio.io/istio/pkg/test/framework/components/pilot"
+)
+
+type testPolicy struct {
+	t         *testing.T
+	env       *kube.Environment
+	namespace string
+	name      string
+}
+
+func (p testPolicy) TearDown() {
+	p.t.Logf("Tearing down policy %q.", p.name)
+	if err := p.env.Delete(p.namespace, path.Join("testdata", p.name)); err != nil {
+		p.t.Fatalf("Cannot delete %q from namespace %q: %v", p.name, p.namespace, err)
+	}
+}
+
+func setupPolicy(t *testing.T, env *kube.Environment, namespace string, name string) *testPolicy {
+	t.Logf("Applying policies from %q.", name)
+	if err := env.Apply(namespace, path.Join("testdata", name)); err != nil {
+		t.Fatalf("Cannot apply %q to namespace %q: %v", name, namespace, err)
+		return nil
+	}
+	return &testPolicy{
+		t:         t,
+		env:       env,
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+type connection struct {
+	from            apps.KubeApp
+	to              apps.KubeApp
+	protocol        apps.AppProtocol
+	port            int
+	expectedSuccess bool
+}
+
+func checkConnection(conn connection) error {
+	ep := conn.to.EndpointForPort(conn.port)
+	if ep == nil {
+		return fmt.Errorf("cannot get upstream endpoint for connection test %v", conn)
+	}
+
+	results, err := conn.from.Call(ep, apps.AppCallOptions{Protocol: conn.protocol})
+	if conn.expectedSuccess {
+		if err != nil || len(results) == 0 || results[0].Code != "200" {
+			// Addition log for debugging purpose.
+			if err != nil {
+				fmt.Printf("Error: %#v\n", err)
+			} else if len(results) == 0 {
+				fmt.Printf("No result\n")
+			} else {
+				fmt.Printf("Result: %v\n", results[0])
+			}
+			return fmt.Errorf("%s to %s:%d using %s: expected success, actually failed",
+				conn.from.Name(), conn.to.Name(), conn.port, conn.protocol)
+		}
+	} else {
+		if err == nil && len(results) > 0 && results[0].Code == "200" {
+			return fmt.Errorf("%s to %s:%d using %s: expected failed, actually success",
+				conn.from.Name(), conn.to.Name(), conn.port, conn.protocol)
+		}
+	}
+	return nil
+}
+
+const (
+	defaultRetryBudget = 10
+	retryDelay         = time.Second
+)
+
+// TODO: move this to framework.
+func runRetriableTest(t *testing.T, testName string, f func() error) {
```

Comment:
There's already a retry function .... see pkg/test/util/retry

Reply:
Thanks for the pointer. Changed to use that one.

Reply:
Ping. Is this good to go?